src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -0,0 +1,842 @@
+        /// Constructs a new <see cref="Utf8JsonWriter"/> instance with a specified <paramref name="bufferWriter"/>.
+        /// </summary>
+        /// <param name="bufferWriter">An instance of <see cref="IBufferWriter{T}" /> used as a destination for writing JSON text into.</param>
+        /// <param name="state">If this is the first call to the ctor, pass in a default state. Otherwise,
+        /// capture the state from the previous instance of the <see cref="Utf8JsonWriter"/> and pass that back.</param>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown when the instance of <see cref="IBufferWriter{T}" /> that is passed in is null.
+        /// </exception>
+        /// <remarks>
+        /// Since this type is a ref struct, it is a stack-only type and all the limitations of ref structs apply to it.
+        /// This is the reason why the ctor accepts a <see cref="JsonWriterState"/>.
+        /// </remarks>
+        public Utf8JsonWriter(IBufferWriter<byte> bufferWriter, JsonWriterState state = default)
+        {
+            _output = bufferWriter ?? throw new ArgumentNullException(nameof(bufferWriter));
+            _buffered = 0;
+            BytesCommitted = 0;
+            _buffer = _output.GetSpan();
+
+            _inObject = state._inObject;
+            _isNotPrimitive = state._isNotPrimitive;
+            _tokenType = state._tokenType;
+            _writerOptions = state._writerOptions;
+            _bitStack = state._bitStack;
+
+            _currentDepth = state._currentDepth;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void Advance(int count)
+        {
+            Debug.Assert(count >= 0 && _buffered <= int.MaxValue - count);
+
+            _buffered += count;
+            _buffer = _buffer.Slice(count);
+        }
+
+        /// <summary>
+        /// Advances the underlying <see cref="IBufferWriter{T}" /> based on what has been written so far.
+        /// </summary>
+        /// <param name="isFinalBlock">Let's the writer know whether more data will be written. This is used to validate
+        /// that the JSON written sor far is structurally valid if no more data is to follow.</param>
+        /// <exception cref="JsonWriterException">
+        /// Thrown when incomplete JSON has been written and <paramref name="isFinalBlock"/> is true.
+        /// (for example when an open object or array needs to be closed).
+        /// </exception>
+        public void Flush(bool isFinalBlock = true)
+        {
+            if (isFinalBlock && !_writerOptions.SkipValidation && CurrentDepth != 0)
+                ThrowHelper.ThrowJsonWriterException(ExceptionResource.ZeroDepthAtEnd, _currentDepth);
+
+            Flush();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void Flush()
+        {
+            BytesCommitted += _buffered;
+            _output.Advance(_buffered);
+            _buffered = 0;
```

Comment:
Are these the correct order of operations if Advance throws?  Seems like _output.Advance should be first?